### PR TITLE
Improve drawing of annotations with matplotlib

### PR DESCRIPTION
**Pull request by mscheltienne**
Looks like I'm going again for an intertwined PR with https://github.com/mne-tools/mne-qt-browser/pull/178 :worried: 
Changes:

- [x] Use blitting for faster render in matplotlib. This [TODO item](https://github.com/mne-tools/mne-python/blob/0d4550b77f2e546f37e69ba4c709cb71132b1848/mne/viz/_mpl_figure.py#L1141-L1142) seems outdated. It's been in since 0.22 (I think?) and I don't see any additional bug compared to `useblit=False` anymore. Maybe fixed upstream? The improvement is drastic.
- [ ] Fix wrapping of annotations around when you exit the view/plot area. The issue is present both with or without blitting, but it's emphasized with blitting.

```
import numpy as np
from mne import create_info
from mne.io import RawArray
from mne.viz import set_browser_backend


set_browser_backend("matplotlib")
data = np.random.randn(3, 100000)
raw = RawArray(data, create_info(3, 1000, "eeg"))
raw.plot(theme="light", scalings="auto")
```

On the screencast, I move my mouse on the channel scrollbar on the right, which yields this bug.

[Screencast from 08-02-2023 11:20:31 AM.webm](https://github.com/mne-tools/mne-python/assets/73893616/a58701cd-6ee1-44a7-a63c-23ffd532833e)

- [ ] Fix merge of dragged annotations. Annotations can not be dragged in the matplotlib backend while it is possible in the Qt backend. The [function](https://github.com/mne-tools/mne-python/blob/23542f5f1c6ebcc52c50471a0dbede85cd04a7b2/mne/viz/utils.py#L1854-L1868) used to merge annotations does not support merge of dragged annotations, only of newly created annotations which overlap with another or of resized annotations which yields an overlap. It lacks knowledge of which annotation was modified within `mne.inst.annotations` and doesn't support changing both boundaries at once. 
- [ ] (optional) Enable dragging of annotations in the maptlotlib backend
- [x] Fix warning emitted when no annotation-label exists. Before, a traceback would be yielded because `buttons` was set to None, and thus it doesn't have the attribute `labels`.
- [ ] (optional) Disable the span selector when no annotation-label exists.

EDIT:
- [ ] Fix addition of more than one label. For now, it raises:

```

Traceback (most recent call last):
  File "/home/scheltie/pyvenv/mscheltienne/mne-python/lib/python3.10/site-packages/matplotlib/cbook/__init__.py", line 309, in process
    func(*args, **kwargs)
  File "/home/scheltie/pyvenv/mscheltienne/mne-python/lib/python3.10/site-packages/matplotlib/widgets.py", line 1114, in _clear
    self.ax.draw_artist(self._checks)
  File "/home/scheltie/pyvenv/mscheltienne/mne-python/lib/python3.10/site-packages/matplotlib/axes/_base.py", line 3074, in draw_artist
    a.draw(self.figure.canvas.get_renderer())
  File "/home/scheltie/pyvenv/mscheltienne/mne-python/lib/python3.10/site-packages/matplotlib/artist.py", line 72, in draw_wrapper
    return draw(artist, renderer)
  File "/home/scheltie/pyvenv/mscheltienne/mne-python/lib/python3.10/site-packages/matplotlib/collections.py", line 971, in draw
    self.set_sizes(self._sizes, self.figure.dpi)
AttributeError: 'NoneType' object has no attribute 'dpi'
```

Might be related to https://github.com/matplotlib/matplotlib/issues/21569

[Screencast from 08-02-2023 01:13:29 PM.webm](https://github.com/mne-tools/mne-python/assets/73893616/927a5e1b-1143-4554-bfa7-5196c0fa5d70)

### CODE REVIEW

File: mne/viz/tests/test_raw.py [hunk 1]
```diff
@@ -827,6 +827,87 @@ def test_remove_annotations(raw, hide_which, browser_backend):
     assert len(raw.annotations) == len(hide_which)
 
 
+def test_merge_annotations(raw, browser_backend):
+    """Test merging of annotations in the Qt backend.
+
+    Let's not bother in figuring out on which sample the _fake_click actually
+    dropped the annotation, especially with the 600.614 Hz weird sampling rate.
+    -> atol = 10 / raw.info["sfreq"]
```
Comment (mscheltienne) [inline mne/viz/tests/test_raw.py:835]: If you have a better idea to get that test working, I'm listening. In https://github.com/mne-tools/mne-qt-browser/pull/178, it failed the Ubuntu and Windows CIs ([link](https://github.com/mne-tools/mne-qt-browser/actions/runs/5751433788/job/15590173072)) with `atol = 2 / raw.info["sfreq"]`. Locally, on an ubuntu-based distro it was passing..

It looks to me like the `_fake_click` location is not very precise. I tried changing the onset/duration/click locations with values corresponding to an exact sample, without luck. For instance, I tried to click on the right edge of an annotation and it was working.. within +/- 15 samples of the actual edge location.

**Comment by mscheltienne** [comment]
In this PR:

- Fixes warning emitted when no annotation-label exist
- Refactor `_update_annotation_segments` to avoid the for loop
- Apply pending TODO item (req. `matplotlib` > 3.0) for inline bold formatting of the instructions
- Enable blitting for faster render
- Add a test for https://github.com/mne-tools/mne-qt-browser/pull/178

I did not add a "Draggable" behavior for annotations to the MPL backend. It looks to me like a lot of work and would require something similar to the `DraggableLine` defined [here](https://github.com/mne-tools/mne-python/blob/69fbf81a4594acc5b42c024e9c9ead954c9ab443/mne/viz/utils.py#L1902-L1968) but for the fill area defined [here](https://github.com/mne-tools/mne-python/blob/69fbf81a4594acc5b42c024e9c9ead954c9ab443/mne/viz/_mpl_figure.py#L1431).

I will re-raise the 2 other bugs (traceback yielded when adding labels + wrapping around when exiting the main ax area) in other issues. I don't know how to fix them at this time, and neither is blocking for this PR.

**Comment by mscheltienne** [comment]
And CIs should be failing here due to the added test which can not pass without https://github.com/mne-tools/mne-qt-browser/pull/178

**Review by larsoner — APPROVED** [review]
@drammock I'll let you merge if happy

@mscheltienne just double-checking -- the top comment has this un-checked box:

> Fix wrapping of annotations around when you exit the view/plot area. The issue is present both with or without blitting, but it's emphasized with blitting.

Should we wait to merge or is it good to go from your end?

**Comment by mscheltienne** [comment]
I have no clue how to fix it for now, so I will re-raise it in a separate issue to keep track. But it's not blocking at the moment.

**Comment by drammock** [comment]
the macOS ARM failure looks vaguely related (to other recent PRs at least):

```
mne/viz/tests/test_raw.py:857: in test_merge_annotations
    assert len(raw.annotations) == 5
E   assert 6 == 5
E    +  where 6 = len(<Annotations | 6 segments: bad_test (3), test (3)>)
E    +    where <Annotations | 6 segments: bad_test (3), test (3)> = <Raw | test_raw.fif, 9 x 14400 (24.0 s), ~3.9 MB, data loaded>.annotations
```

is it an issue of the CI config not using the right version of mne-qt-browser?

**Comment by mscheltienne** [comment]
Yes, it's not running `mne-qt-browser` main: 

```
└☑ mne-qt-browser       0.5.2
        /opt/homebrew/lib/python3.10/site-packages/mne_qt_browser
```

**Comment by drammock** [comment]
> We could mark the test as xfail for qt-browser version below 0.5.3.

let's do that... I don't know how long it will take for #11859 to resolve and I don't want Cirrus failure on every PR between now and then.

**Comment by larsoner** [comment]
> let's do that... I don't know how long it will take for https://github.com/mne-tools/mne-python/pull/11859 to resolve and I don't want Cirrus failure on every PR between now and then.

That PR is very close to done... but even then it would be good to have that `xfail` so that other devs running tests who haven't updated to `main` of `mne-qt-browser` can still have things pass.

**Comment by larsoner** [comment]
@drammock should come back green now 🤞

File: mne/viz/tests/test_raw.py (hunk 1 above)
Comment (larsoner) [reply mne/viz/tests/test_raw.py:835]: > It looks to me like the _fake_click location is not very precise

Yes it's possible and I think it's okay to have a loose tolerance. As long as it's strict enough to catch if we broke things it should be fine. We can always improve the tolerance later

**Comment by mscheltienne** [comment]
Thank you @larsoner for the CIs fixes!

File: mne/viz/tests/test_raw.py [hunk 2]
```diff
@@ -827,6 +827,89 @@ def test_remove_annotations(raw, hide_which, browser_backend):
...
+    """
+    if browser_backend.name == "matplotlib":
+        pytest.skip("The MPL backend does not support draggable annotations.")
+    elif not check_version("mne_qt_browser", "0.5.3"):
+        pytest.xfail("mne_qt_browser < 0.5.3 does not merge properly")
```
Comment (drammock) [inline mne/viz/tests/test_raw.py:840]:
```suggestion
        pytest.xfail("mne_qt_browser < 0.5.3 does not merge annotations properly")
```

**Review by drammock — APPROVED** [review]
the test is not very DRY but I'm not sure it would actually be easier to follow if it were DRY. +1 for merge after 1 tiny wording change to the xfail message

**Pull request merged**